listenbrainz/webserver/static/js/src/CBReviewModal.tsx

```diff
@@ -0,0 +1,703 @@
+        entity_id: entityToReview.mbid,
+        entity_type: entityToReview.type,
+        text: textContent,
+        languageCode: language,
+        rating: nonZeroRating,
+      };
+
+      try {
+        const response = await APIService.submitReviewToCB(
+          access_token,
+          reviewToSubmit
+        );
+        if (response) {
+          newAlert(
+            "success",
+            `Your review was submitted to CritiqueBrainz!`,
+            `${listen.track_metadata.artist_name} - ${entityToReview?.name}`
+          );
+          this.setState({
+            textContent: "",
+            rating: 0,
+            reviewValidateAlert: null,
+            success: true,
+            reviewMBID: response.id,
+          });
+        }
+      } catch (error) {
+        if (maxRetries > 0 && error.message === "invalid_token") {
+          /* Need to refresh token and retry with new token */
+          const newToken = await this.refreshCritiquebrainzToken();
+          this.submitReviewToCB(newToken, maxRetries - 1);
+        } else
+          this.handleError(
+            error,
+            "Error while submitting review to CritiqueBrainz"
+          );
+      }
+      this.setState({ loading: false });
+    }
+  };
+
+  getModalBody = (hasPermissions: boolean) => {
+    const { listen } = this.props;
+    const {
+      entityToReview,
+      textContent,
+      rating,
+      recordingEntity,
+      artistEntity,
+      releaseGroupEntity,
+      acceptLicense,
+      reviewValidateAlert,
+      language,
+      success,
+      reviewMBID,
+    } = this.state;
+
+    /* User hasn't logged into CB yet, prompt them to authenticate */
+    if (!hasPermissions) {
+      return (
+        <div>
+          Before you can submit reviews for your Listens to{" "}
+          <a href={this.CBBaseUrl}>CritiqueBrainz</a>, you must{" "}
+          <b> connect to your CritiqueBrainz </b> account from ListenBrainz.
+          {this.CBInfoButton}
+          <br />
+          <br />
+          You can connect to your CritiqueBrainz account by visiting the
+          <a href={`${window.location.origin}/profile/music-services/details/`}>
+            {" "}
+            music services page.
+          </a>
+        </div>
+      );
+    }
+
+    /* Success message */
+    if (success && entityToReview) {
+      return (
+        <div>
+          Thanks for submitting your review for <b>{entityToReview.name}</b>!
+          <br />
+          <br />
+          You can access your CritiqueBrainz review by clicking{" "}
+          <a href={`${this.CBBaseUrl}/${reviewMBID}`}> here.</a>
+        </div>
+      );
+    }
+
+    /* None of the three entities were found for the Listen */
+    if (!entityToReview) {
+      return (
+        <div>
+          We could not find any reviewable entities associated with{" "}
+          <b>{listen.track_metadata?.track_name}</b> by{" "}
+          <b>{listen.track_metadata?.artist_name}.</b> :(
```

Comment:
I see this as a good opportunity to add missing data to MusicBrainz, if we couldn't find anything.

```suggestion
          We could not link <b>{listen.track_metadata?.track_name}</b> by{" "}
          <b>{listen.track_metadata?.artist_name}.</b> to any recording or artist on
          MusicBrainz. If you can't find them when searching <a href="https://musicbrainz.org/search">on MusicBrainz</a> either,
          please consider <a href="https://musicbrainz.org/doc/Introduction_to_Editing">adding them to our database</a>.
```